File: duplex/go/wire_test.go
package duplex_test

import (
	"encoding/json"
	"errors"
	bitwire "github.com/Bitspark/bitwire/wire/go"
	"reflect"
	"sync"
	"testing"

	"github.com/Bitspark/nightseam/duplex/go"
)

func TestPathEncodingIsCanonicalAndComposable(t *testing.T) {
	paths := [][]string{{}, {""}, {"a", "b"}, {"a.b"}, {"a", "b:c"}, {"é", "e\u0301", "😀", "\ufeff", "\x00"}}
	seen := map[string]bool{}
	for _, path := range paths {
		encoded, err := duplex.EncodePath(path)
		if err != nil {
			t.Fatal(err)
		}
		if seen[encoded] {
			t.Fatalf("paths alias at %q", encoded)
		}
		seen[encoded] = true
		decoded, err := duplex.DecodePath(encoded)
		if err != nil || !reflect.DeepEqual(decoded, path) {
			t.Fatalf("%q: %#v, %v", encoded, decoded, err)
		}
		for _, suffix := range paths {
			b, _ := duplex.EncodePath(suffix)
			combined, _ := duplex.EncodePath(append(append([]string{}, path...), suffix...))
			if combined != encoded+b {
				t.Fatal("prefixing did not compose by concatenation")
			}
		}
	}
	if got, _ := duplex.EncodePath([]string{"a", "😀", ""}); got != "1:a4:😀0:" {
		t.Fatal(got)
	}
	for _, malformed := range []string{"01:a", "00:", "1", ":", "-1:a", "2:a", "1:é", "99999999999999999999999999999:x", "1:\xff"} {
		if _, err := duplex.DecodePath(malformed); err == nil {
			t.Fatalf("accepted %q", malformed)
		}
	}
	if _, err := duplex.EncodePath([]string{"\xff"}); err == nil {
		t.Fatal("accepted non-scalar UTF-8")
	}
}

// queuedRoot is a deterministic endpoint fixture. Only drain executes queued
// deliveries, so composition cannot pass the asynchronous check by timing luck.
type queuedRoot struct {
	mu      sync.Mutex
	queue   []queuedDelivery
	current *rootAttachment
	closed  bool
	closes  int
}
type rootAttachment struct{ receiver bitwire.Receiver }
type queuedDelivery struct {
	path    []string
	message bitwire.Message
}
type nonComparableRoot struct {
	*queuedRoot
	marker []int
}

func newRoot() *queuedRoot { return &queuedRoot{} }
func (r *queuedRoot) Send(path []string, message bitwire.Message) error {
	if _, err := duplex.EncodePath(path); err != nil {
		return err
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.closed {
		return duplex.ErrClosed
	}
	r.queue = append(r.queue, queuedDelivery{append([]string{}, path...), message})
	return nil
}
func (r *queuedRoot) Receive(receiver bitwire.Receiver) (func(), error) {
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.closed {
		return nil, duplex.ErrClosed
	}
	if r.current != nil {
		return nil, duplex.ErrReceiverExists
	}
	attachment := &rootAttachment{receiver}
	r.current = attachment
	return func() {
		r.mu.Lock()
		if r.current == attachment {
			r.current = nil
		}
		r.mu.Unlock()
	}, nil
}
func (r *queuedRoot) attached() bool {
	r.mu.Lock()
	defer r.mu.Unlock()
	return r.current != nil
}
func (r *queuedRoot) captured() bitwire.Receiver {
	r.mu.Lock()
	defer r.mu.Unlock()
	return r.current.receiver
}
func (r *queuedRoot) Close(code bitwire.Code, reason string) error {
	r.mu.Lock()
	if r.closed {
		r.mu.Unlock()
		return nil
	}
	r.closed = true
	r.closes++
	attachment := r.current
	r.current = nil
	r.mu.Unlock()
	if attachment != nil && attachment.receiver.Closed != nil {
		attachment.receiver.Closed(code, reason)
	}
	return nil
}
func (r *queuedRoot) drain() {
	for {
		r.mu.Lock()
		if len(r.queue) == 0 {
			r.mu.Unlock()
			return
		}
		d := r.queue[0]
		r.queue = r.queue[1:]
		attachment := r.current
		r.mu.Unlock()
		if attachment != nil && attachment.receiver.Message != nil {
			attachment.receiver.Message(d.path, d.message)
		}
	}
}

type sendOnly func([]string, bitwire.Message) error

func (s sendOnly) Send(path []string, message bitwire.Message) error { return s(path, message) }

func TestWireSelectionsGrantOnlySendAccess(t *testing.T) {
	root := newRoot()
	prefix := []string{"a.b"}
	selected := duplex.At(sendOnly(root.Send), prefix)
	prefix[0] = "changed"
	for _, view := range []bitwire.Wire{selected, duplex.At(root, nil), duplex.At(selected, []string{"😀"})} {
		if _, ok := view.(interface {
			Receive(bitwire.Receiver) (func(), error)
		}); ok {
			t.Fatal("selection grants receive authority")
		}
		if _, ok := view.(interface {
			Close(bitwire.Code, string) error
		}); ok {
			t.Fatal("selection grants lifecycle authority")
		}
	}
	path := []string{"call"}
	if err := duplex.At(selected, []string{"😀"}).Send(path, bitwire.Message{}); err != nil {
		t.Fatal(err)
	}
	path[0] = "changed"
	if !reflect.DeepEqual(root.queue[0].path, []string{"a.b", "😀", "call"}) {
		t.Fatal(root.queue)
	}
}

func TestMountPreservesPathsFramesAndReturnCapability(t *testing.T) {
	left, right, reply := newRoot(), newRoot(), newRoot()
	children := map[string]bitwire.Endpoint{"left": left, "": right}
	mounted := duplex.Mount(children)
	children["left"] = reply
	address := &bitwire.ReturnAddress{Wire: nonComparableRoot{reply, []int{1}}}
	var paths [][]string
	var received []bitwire.Message
	_, err := mounted.Receive(bitwire.Receiver{Message: func(path []string, message bitwire.Message) {
		paths = append(paths, path)
		received = append(received, message)
	}})
	if err != nil {
		t.Fatal(err)
	}
	frames := []bitwire.ProfileFrame{
		{Version: 1, Kind: bitwire.ProfileRequest, ID: "c:1", Params: json.RawMessage(`{"n":9007199254740993}`), Meta: map[string]string{"tag": "value"}},
		{Version: 1, Kind: bitwire.ProfileResponse, ID: "c:1", Error: &bitwire.ProfileError{Code: "refused", Message: "No", Data: json.RawMessage(`{"why":"test"}`)}},
		{Version: 1, Kind: bitwire.ProfileEvent, Data: json.RawMessage(`null`)},
		{Version: 1, Kind: bitwire.ProfileCancel, ID: "c:1"},
	}
	view := duplex.At(duplex.At(mounted, []string{"left"}), []string{"😀"})
	for _, frame := range frames {
		if err := view.Send([]string{"call"}, bitwire.Message{Frame: frame, Return: address}); err != nil {
			t.Fatal(err)
		}
	}
	if len(received) != 0 || len(left.queue) != 4 || len(reply.queue) != 0 {
		t.Fatal("composition changed dispatch ownership")
	}
	for _, delivery := range left.queue {
		if !reflect.DeepEqual(delivery.path, []string{"😀", "call"}) {
			t.Fatal(delivery.path)
		}
	}
	left.drain()
	for i, frame := range frames {
		if !reflect.DeepEqual(paths[i], []string{"left", "😀", "call"}) || !reflect.DeepEqual(received[i].Frame, frame) || received[i].Return != address {
			t.Fatal(paths[i], received[i])
		}
	}
	if err := mounted.Send([]string{""}, bitwire.Message{}); err != nil {
		t.Fatal(err)
	}
	right.drain()
	if !reflect.DeepEqual(paths[4], []string{""}) {
		t.Fatal(paths[4])
	}
	for _, path := range [][]string{nil, {"missing"}} {
		if err := mounted.Send(path, bitwire.Message{}); !errors.Is(err, duplex.ErrNoRoute) {
			t.Fatal(err)
		}
	}
	if err := mounted.Send([]string{"left", "\xff"}, bitwire.Message{}); !errors.Is(err, duplex.ErrPath) {
		t.Fatal(err)
	}
}

func TestMountRefusesDuplicateAttachmentAndRebindsWithoutStealingCapturedDeliveries(t *testing.T) {
	root, reply := newRoot(), newRoot()
	mounted := duplex.Mount(map[string]bitwire.Endpoint{"service": root})
	address := &bitwire.ReturnAddress{Wire: duplex.At(reply, nil)}
	var old, fresh []bitwire.ProfileKind
	var oldPaths [][]string
	closed := 0
	detach, err := mounted.Receive(bitwire.Receiver{
		Message: func(path []string, message bitwire.Message) {
			oldPaths = append(oldPaths, path)
			old = append(old, message.Frame.Kind)
			if message.Return != address {
				t.Fatal("return capability changed")
			}
		},
		Closed: func(bitwire.Code, string) { closed++ },
	})
	if err != nil {
		t.Fatal(err)
	}
	captured := root.captured()
	if _, err := mounted.Receive(bitwire.Receiver{}); !errors.Is(err, duplex.ErrReceiverExists) {
		t.Fatal(err)
	}
	captured.Message([]string{"wait"}, bitwire.Message{Frame: bitwire.ProfileFrame{Kind: bitwire.ProfileRequest}, Return: address})
	detach()
	detach()
	if root.attached() || root.closed {
		t.Fatal("detach retained ownership or closed borrowed root")
	}
	freshDetach, err := mounted.Receive(bitwire.Receiver{Message: func(_ []string, message bitwire.Message) { fresh = append(fresh, message.Frame.Kind) }})
	if err != nil {
		t.Fatal(err)
	}
	detach() // The old token must never remove the new attachment.
	captured.Closed(duplex.CodeNormal, "stale close")
	captured.Message([]string{"wait"}, bitwire.Message{Frame: bitwire.ProfileFrame{Kind: bitwire.ProfileCancel}, Return: address})
	if err := address.Wire.Send(nil, bitwire.Message{Frame: bitwire.ProfileFrame{Kind: bitwire.ProfileResponse}}); err != nil {
		t.Fatal(err)
	}
	if err := mounted.Send([]string{"service", "new"}, bitwire.Message{Frame: bitwire.ProfileFrame{Kind: bitwire.ProfileEvent}}); err != nil {
		t.Fatal(err)
	}
	root.drain()
	if !reflect.DeepEqual(old, []bitwire.ProfileKind{bitwire.ProfileRequest, bitwire.ProfileCancel}) || !reflect.DeepEqual(fresh, []bitwire.ProfileKind{bitwire.ProfileEvent}) || closed != 0 || len(reply.queue) != 1 {
		t.Fatal(old, fresh, closed, reply.queue)
	}
	if !reflect.DeepEqual(oldPaths, [][]string{{"service", "wait"}, {"service", "wait"}}) {
		t.Fatal(oldPaths)
	}
	freshDetach()
	_ = mounted.Close(duplex.CodeNormal, "done")
	if closed != 0 || root.closed {
		t.Fatal("detached owner or borrowed child was closed")
	}
}

func TestMountAttachmentFailureRollsBackOnlyItsBorrowedAttachments(t *testing.T) {
	for _, duplicate := range []bool{false, true} {
		t.Run(map[bool]string{false: "occupied-child", true: "aliased-child"}[duplicate], func(t *testing.T) {
			first, occupied := newRoot(), newRoot()
			if duplicate {
				occupied = first
			} else {
				if _, err := occupied.Receive(bitwire.Receiver{}); err != nil {
					t.Fatal(err)
				}
			}
			mounted := duplex.Mount(map[string]bitwire.Endpoint{"a": first, "z": occupied})
			if _, err := mounted.Receive(bitwire.Receiver{}); !errors.Is(err, duplex.ErrReceiverExists) {
				t.Fatal(err)
			}
			if first.attached() || first.closed || occupied.closed {
				t.Fatal("failed acquisition leaked or closed a child")
			}
			if !duplicate && !occupied.attached() {
				t.Fatal("rollback removed another owner")
			}
			if _, err := first.Receive(bitwire.Receiver{}); err != nil {
				t.Fatal(err)
			}
		})
	}
}

func TestMountedChildEndKeepsHealthySiblingAndEndsOwnerOnce(t *testing.T) {
	left, right := newRoot(), newRoot()
	mounted := duplex.Mount(map[string]bitwire.Endpoint{"left": left, "right": right})
	closed, deliveries := 0, 0
	_, err := mounted.Receive(bitwire.Receiver{
		Message: func(path []string, _ bitwire.Message) {
			if !reflect.DeepEqual(path, []string{"right", "call"}) {
				t.Fatal(path)
			}
			deliveries++
		},
		Closed: func(code bitwire.Code, reason string) {
			closed++
			if code != duplex.CodeNormal || reason != "last" {
				t.Fatal(code, reason)
			}
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	stale := left.captured()
	_ = left.Close(duplex.CodeNormal, "first")
	stale.Closed(duplex.CodeNormal, "duplicate")
	if closed != 0 {
		t.Fatal("one child ended the mount attachment")
	}
	if err := mounted.Send([]string{"right", "call"}, bitwire.Message{}); err != nil {
		t.Fatal(err)
	}
	right.drain()
	_ = right.Close(duplex.CodeNormal, "last")
	if closed != 1 || deliveries != 1 {
		t.Fatal(closed, deliveries)
	}
	if err := mounted.Send(nil, bitwire.Message{}); !errors.Is(err, duplex.ErrNoRoute) {
		t.Fatal("child ending permanently closed mount", err)
	}
	if _, err := mounted.Receive(bitwire.Receiver{}); !errors.Is(err, duplex.ErrClosed) {
		t.Fatal(err)
	}
	_ = mounted.Close(duplex.CodeNormal, "mount")
	if closed != 1 {
		t.Fatal(closed)
	}
}

func TestMountCloseDetachesOwnAttachmentAndPreservesChildren(t *testing.T) {
	root := newRoot()
	mounted := duplex.Mount(map[string]bitwire.Endpoint{"": root})
	closed := 0
	_, err := mounted.Receive(bitwire.Receiver{Closed: func(code bitwire.Code, reason string) {
		closed++
		if code != duplex.CodeNormal || reason != "mount ended" {
			t.Error(code, reason)
		}
		_ = mounted.Close(code, reason)
		if _, err := root.Receive(bitwire.Receiver{}); err != nil {
			t.Error("child was not released before closure callback", err)
		}
	}})
	if err != nil {
		t.Fatal(err)
	}
	_ = mounted.Close(duplex.CodeNormal, "mount ended")
	_ = mounted.Close(duplex.CodeNormal, "again")
	if closed != 1 || root.closes != 0 || !root.attached() {
		t.Fatal(closed, root.closes, root.attached())
	}
	if err := mounted.Send([]string{""}, bitwire.Message{}); !errors.Is(err, duplex.ErrClosed) {
		t.Fatal(err)
	}
	if _, err := mounted.Receive(bitwire.Receiver{}); !errors.Is(err, duplex.ErrClosed) {
		t.Fatal(err)
	}
	if err := root.Send(nil, bitwire.Message{}); err != nil {
		t.Fatal(err)
	}
}

type registeringRoot struct {
	*queuedRoot
	registered chan struct{}
	resume     chan struct{}
}

func (r *registeringRoot) Receive(receiver bitwire.Receiver) (func(), error) {
	detach, err := r.queuedRoot.Receive(receiver)
	close(r.registered)
	<-r.resume
	return detach, err
}
func TestMountCloseDuringReceiveDisposesLateChildAttachment(t *testing.T) {
	root := &registeringRoot{newRoot(), make(chan struct{}), make(chan struct{})}
	mounted := duplex.Mount(map[string]bitwire.Endpoint{"x": root})
	finished := make(chan error, 1)
	closed := 0
	go func() {
		_, err := mounted.Receive(bitwire.Receiver{Closed: func(bitwire.Code, string) { closed++ }})
		finished <- err
	}()
	<-root.registered
	_ = mounted.Close(duplex.CodeNormal, "done")
	close(root.resume)
	if err := <-finished; !errors.Is(err, duplex.ErrClosed) {
		t.Fatal(err)
	}
	if root.attached() || root.closes != 0 || closed != 1 {
		t.Fatal(root.attached(), root.closes, closed)
	}
}

type endingRoot struct{ *queuedRoot }

func (r *endingRoot) Receive(receiver bitwire.Receiver) (func(), error) {
	detach, err := r.queuedRoot.Receive(receiver)
	if err == nil {
		_ = r.Close(duplex.CodeNormal, "ended during acquisition")
	}
	return detach, err
}
func TestMountChildEndingDuringAcquisitionRollsBackHealthySibling(t *testing.T) {
	healthy := newRoot()
	ending := &endingRoot{newRoot()}
	mounted := duplex.Mount(map[string]bitwire.Endpoint{"a": healthy, "z": ending})
	if _, err := mounted.Receive(bitwire.Receiver{}); !errors.Is(err, duplex.ErrClosed) {
		t.Fatal(err)
	}
	if healthy.attached() || ending.attached() || healthy.closed {
		t.Fatal("partial acquisition retained a child")
	}
	if _, err := healthy.Receive(bitwire.Receiver{}); err != nil {
		t.Fatal(err)
	}
}

func TestEmptyMountStillOwnsOneDetachableAttachment(t *testing.T) {
	mounted := duplex.Mount(nil)
	detach, err := mounted.Receive(bitwire.Receiver{})
	if err != nil {
		t.Fatal(err)
	}
	if _, err := mounted.Receive(bitwire.Receiver{}); !errors.Is(err, duplex.ErrReceiverExists) {
		t.Fatal(err)
	}
	detach()
	closed := 0
	_, err = mounted.Receive(bitwire.Receiver{Closed: func(bitwire.Code, string) { closed++ }})
	if err != nil {
		t.Fatal(err)
	}
	detach()
	_ = mounted.Close(duplex.CodeNormal, "done")
	if closed != 1 {
		t.Fatal(closed)
	}
}

File: duplex/ts/src/wire.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';
import { at, mount, encodePath, decodePath, WireError } from './wire.ts';
import type { Wire, Endpoint, Message, Receiver, Path, ProfileFrame } from '@bitspark/bitwire';

test('path encoding is canonical, injective and composes by concatenation', () => {
  const paths: Path[] = [[], [''], ['a', 'b'], ['a.b'], ['a', 'b:c'], ['é', 'e\u0301', '😀', '\ufeff', '\0']];
  const seen = new Set<string>();
  for (const path of paths) {
    const encoded = encodePath(path);
    assert.equal(seen.has(encoded), false);
    seen.add(encoded);
    assert.deepEqual(decodePath(encoded), path);
    for (const suffix of paths) assert.equal(encodePath([...path, ...suffix]), encoded + encodePath(suffix));
  }
  assert.equal(encodePath(['a', '😀', '']), '1:a4:😀0:');
  for (const malformed of [
    '01:a',
    '00:',
    '1',
    ':',
    '-1:a',
    '2:a',
    '1:é',
    '99999999999999999999999999999:x',
    '1:\ud800',
  ])
    assert.throws(() => decodePath(malformed));
  assert.throws(() => encodePath(['\ud800']));
  assert.throws(() => encodePath(['\udc00']));
});

// A controlled single-attachment fixture: no scheduler or physical-carrier claim.
class QueuedEndpoint implements Endpoint {
  readonly queue: { path: Path; message: Message }[] = [];
  receiver: Receiver | undefined;
  attachments = 0;
  detaches = 0;
  closes = 0;
  send(path: Path, message: Message): void {
    if (this.closes) throw new WireError('closed');
    encodePath(path);
    this.queue.push({ path: [...path], message });
  }
  receive(receiver: Receiver): () => void {
    if (this.closes) throw new WireError('closed');
    if (this.receiver) throw new WireError('receiver_exists');
    this.attachments++;
    this.receiver = receiver;
    let active = true;
    return () => {
      if (!active) return;
      active = false;
      this.detaches++;
      if (this.receiver === receiver) this.receiver = undefined;
    };
  }
  close(code = 1000, reason = ''): void {
    if (this.closes) return;
    this.closes++;
    const receiver = this.receiver;
    this.receiver = undefined;
    receiver?.closed?.(code, reason);
  }
  async drain(): Promise<void> {
    for (let entry = this.queue.shift(); entry; entry = this.queue.shift()) {
      await this.receiver?.message?.(entry.path, entry.message);
    }
  }
}

const event: Message = { frame: { version: 1, kind: 'event', data: null } };

test('selection is send-only, copies prefixes and delegates nested and empty paths without dispatching', () => {
  const root = new QueuedEndpoint();
  let delivered = 0;
  root.receive({
    message: () => {
      delivered++;
    },
  });
  const prefix = ['a.b'];
  const selected = at(root, prefix);
  prefix[0] = 'changed';
  const nested = at(selected, ['😀']);
  const suffix = ['call'];
  nested.send(suffix, event);
  suffix[0] = 'changed';
  at(root, []).send([], event);
  at(at(root, ['a.b']), ['😀']).send(['call'], event);
  assert.equal(delivered, 0);
  assert.deepEqual(
    root.queue.map((entry) => entry.path),
    [['a.b', '😀', 'call'], [], ['a.b', '😀', 'call']],
  );
  for (const entry of root.queue) assert.equal(entry.message, event);
  for (const view of [selected, nested, at(root, [])]) {
    assert.deepEqual(Object.keys(view), ['send']);
    assert.equal('receive' in view, false);
    assert.equal('close' in view, false);
  }
  // A send-only implementation, not just an Endpoint narrowed statically.
  const access: Wire = { send: (path, message) => root.send(path, message) };
  at(access, ['']).send([], event);
  assert.deepEqual(root.queue.at(-1)?.path, ['']);
});

test('one mount receiver spans children and preserves whole messages, return identity and queued dispatch', async () => {
  const left = new QueuedEndpoint(),
    right = new QueuedEndpoint(),
    replacement = new QueuedEndpoint();
  const children = new Map<string, Endpoint>([
    ['left', left],
    ['', right],
  ]);
  const mounted = mount(children);
  children.set('left', replacement);
  const received: { path: Path; message: Message }[] = [];
  const detach = mounted.receive({
    message: (path, message) => {
      received.push({ path, message });
    },
  });
  assert.equal(left.attachments, 1);
  assert.equal(right.attachments, 1);
  const reply: Wire = { send() {} };
  const address = { wire: reply };
  const frames: ProfileFrame[] = [
    { version: 1, kind: 'request', id: 'c:1', params: { n: 42 }, meta: { tag: 'value' } },
    { version: 1, kind: 'response', id: 'c:1', error: { code: 'refused', message: 'No', data: { why: 'test' } } },
    { version: 1, kind: 'event', data: null },
    { version: 1, kind: 'cancel', id: 'c:1' },
  ];
  const messages = frames.map((frame) => ({ frame, return: address }));
  const contexts = new WeakMap<Message, object>();
  for (const message of messages) {
    contexts.set(message, {});
    at(at(mounted, ['left']), ['😀']).send(['call'], message);
  }
  mounted.send([''], event);
  assert.equal(received.length, 0, 'destination ran inside send');
  assert.equal(left.queue.length, 4);
  assert.equal(right.queue.length, 1);
  assert.equal(replacement.queue.length, 0);
  for (const entry of left.queue) assert.deepEqual(entry.path, ['😀', 'call']);
  assert.deepEqual(right.queue[0]?.path, []);
  await left.drain();
  await right.drain();
  assert.equal(received.length, 5);
  for (let i = 0; i < messages.length; i++) {
    const delivered = received[i]!;
    assert.deepEqual(delivered.path, ['left', '😀', 'call']);
    assert.equal(delivered.message, messages[i]);
    assert.equal(delivered.message.return, address);
    assert.equal(contexts.get(delivered.message), contexts.get(messages[i]!));
  }
  assert.deepEqual(received[4], { path: [''], message: event });
  assert.throws(() => mounted.receive({}), { code: 'receiver_exists' });
  assert.equal(left.attachments, 1);
  assert.equal(right.attachments, 1);
  detach();
  assert.equal(left.detaches, 1);
  assert.equal(right.detaches, 1);
});

test('detach allows rebind and a stale disposer cannot detach its successor or notify it', async () => {
  const root = new QueuedEndpoint();
  const mounted = mount(new Map([['service', root]]));
  const calls: string[] = [];
  let closed = 0;
  const first = mounted.receive({
    message: () => {
      calls.push('first');
    },
    closed: () => {
      closed++;
    },
  });
  const captured = root.receiver!;
  mounted.send(['service', 'first'], event);
  await root.drain();
  first();
  const second = mounted.receive({
    message: () => {
      calls.push('second');
    },
    closed: () => {
      closed++;
    },
  });
  first();
  captured.closed?.(1000, 'stale ending');
  mounted.send(['service', 'second'], event);
  await root.drain();
  assert.deepEqual(calls, ['first', 'second']);
  assert.equal(closed, 0);
  assert.equal(root.detaches, 1);
  second();
  second();
  mounted.close();
  assert.equal(root.detaches, 2);
  assert.equal(root.closes, 0);
  assert.equal(closed, 0);
});

test('partial acquisition rolls back earlier children without releasing another owner', () => {
  const left = new QueuedEndpoint(),
    right = new QueuedEndpoint();
  const other: Receiver = {};
  const releaseOther = right.receive(other);
  const mounted = mount(
    new Map([
      ['left', left],
      ['right', right],
    ]),
  );
  let closed = 0;
  assert.throws(
    () =>
      mounted.receive({
        closed: () => {
          closed++;
        },
      }),
    { code: 'receiver_exists' },
  );
  assert.equal(left.receiver, undefined);
  assert.equal(left.detaches, 1);
  assert.equal(right.receiver, other);
  assert.equal(right.detaches, 0);
  assert.equal(closed, 0);
  releaseOther();
  const detach = mounted.receive({});
  detach();
  assert.equal(left.detaches, 2);
  assert.equal(right.detaches, 2);
  assert.equal(left.closes + right.closes, 0);
});

test('two mount keys cannot acquire the same child twice and roll back the first attachment', () => {
  const child = new QueuedEndpoint();
  const mounted = mount(
    new Map([
      ['one', child],
      ['two', child],
    ]),
  );
  assert.throws(() => mounted.receive({}), { code: 'receiver_exists' });
  assert.equal(child.receiver, undefined);
  assert.equal(child.detaches, 1);
  assert.equal(child.closes, 0);
  child.receive({})();
});

test('mount close releases its attachment once and leaves borrowed children usable', async () => {
  const left = new QueuedEndpoint(),
    right = new QueuedEndpoint();
  const mounted = mount(
    new Map([
      ['left', left],
      ['right', right],
    ]),
  );
  let closed = 0;
  const detach = mounted.receive({
    closed: (code, reason) => {
      closed++;
      assert.equal(code, 1000);
      assert.equal(reason, 'mount ended');
      assert.equal(left.receiver, undefined);
      assert.equal(right.receiver, undefined);
      mounted.close();
    },
  });
  const old = left.receiver!;
  mounted.close(1000, 'mount ended');
  mounted.close(1000, 'again');
  detach();
  old.closed?.(1000, 'late child');
  assert.equal(closed, 1);
  assert.equal(left.closes + right.closes, 0);
  assert.equal(left.detaches + right.detaches, 2);
  assert.throws(() => mounted.receive({}), { code: 'closed' });
  assert.throws(() => at(mounted, ['left']).send(['call'], event), { code: 'closed' });
  let received = 0;
  for (const child of [left, right]) {
    const release = child.receive({
      message: () => {
        received++;
      },
    });
    child.send(['call'], event);
    await child.drain();
    release();
  }
  assert.equal(received, 2);
});

test('mount close during child receive releases both early and late returned disposers', () => {
  const left = new QueuedEndpoint(),
    right = new QueuedEndpoint();
  let mounted: Endpoint;
  const child: Endpoint = {
    send: (path, message) => right.send(path, message),
    receive: (receiver) => {
      const detach = right.receive(receiver);
      mounted.close(1000, 'done');
      return detach;
    },
    close: (code, reason) => right.close(code, reason),
  };
  mounted = mount(
    new Map([
      ['left', left],
      ['right', child],
    ]),
  );
  let closed = 0;
  assert.throws(
    () =>
      mounted.receive({
        closed: () => {
          closed++;
        },
      }),
    { code: 'closed' },
  );
  assert.equal(left.receiver, undefined);
  assert.equal(right.receiver, undefined);
  assert.equal(left.detaches, 1);
  assert.equal(right.detaches, 1);
  assert.equal(left.closes + right.closes, 0);
  assert.equal(closed, 1);
});

test('a child ending during acquisition refuses and rolls back the incomplete attachment', () => {
  const left = new QueuedEndpoint(),
    right = new QueuedEndpoint(),
    later = new QueuedEndpoint();
  const child: Endpoint = {
    send: (path, message) => right.send(path, message),
    receive: (receiver) => {
      const detach = right.receive(receiver);
      right.close(1000, 'ended during attachment');
      return detach;
    },
    close: (code, reason) => right.close(code, reason),
  };
  const mounted = mount(
    new Map([
      ['left', left],
      ['right', child],
      ['later', later],
    ]),
  );
  let closed = 0;
  assert.throws(
    () =>
      mounted.receive({
        closed: () => {
          closed++;
        },
      }),
    { code: 'closed' },
  );
  assert.equal(left.receiver, undefined);
  assert.equal(right.receiver, undefined);
  assert.equal(left.detaches, 1);
  assert.equal(right.detaches, 1);
  assert.equal(later.attachments, 0);
  assert.equal(left.closes, 0);
  assert.equal(closed, 0);
  left.receive({})();
});

test('a child ending leaves siblings active and the final child ends only the current attachment', async () => {
  const left = new QueuedEndpoint(),
    right = new QueuedEndpoint();
  const mounted = mount(
    new Map([
      ['left', left],
      ['right', right],
    ]),
  );
  const paths: Path[] = [];
  const endings: { code: number; reason: string }[] = [];
  const detach = mounted.receive({
    message: (path) => {
      paths.push(path);
    },
    closed: (code, reason) => {
      endings.push({ code, reason });
    },
  });
  const firstReceiver = left.receiver!;
  left.close(1000, 'left ended');
  firstReceiver.closed?.(1000, 'duplicate');
  assert.equal(endings.length, 0);
  mounted.send(['right', 'still', 'usable'], event);
  await right.drain();
  assert.deepEqual(paths, [['right', 'still', 'usable']]);
  right.close(1001, 'last ended');
  assert.deepEqual(endings, [{ code: 1001, reason: 'last ended' }]);
  assert.throws(() => mounted.send([], event), { code: 'no_route' });
  detach();
  mounted.close();
  assert.equal(endings.length, 1);
  assert.equal(left.closes, 1);
  assert.equal(right.closes, 1);
  assert.equal(left.detaches, 1);
  assert.equal(right.detaches, 1);
});

test('captured request and cancellation retain the original receiver and async result after detach and rebind', async () => {
  const root = new QueuedEndpoint(),
    reply = new QueuedEndpoint();
  const mounted = mount(new Map([['service', root]]));
  const address = { wire: reply };
  const request: Message = { frame: { version: 1, kind: 'request', id: 'c:1', params: {} }, return: address };
  const cancel: Message = { frame: { version: 1, kind: 'cancel', id: 'c:1' }, return: address };
  const seen: { path: Path; message: Message }[] = [];
  let complete!: () => void;
  const pending = new Promise<void>((resolve) => {
    complete = resolve;
  });
  const detach = mounted.receive({
    message: (path, message) => {
      seen.push({ path, message });
      return pending;
    },
  });
  const captured = root.receiver!;
  assert.equal(captured.message?.(['wait'], request), pending);
  detach();
  let successor = 0;
  const releaseNext = mounted.receive({
    message: () => {
      successor++;
    },
  });
  assert.equal(captured.message?.(['wait'], cancel), pending);
  complete();
  await pending;
  assert.deepEqual(
    seen.map((entry) => entry.path),
    [
      ['service', 'wait'],
      ['service', 'wait'],
    ],
  );
  assert.equal(seen[0]?.message, request);
  assert.equal(seen[1]?.message, cancel);
  assert.equal(seen[1]?.message.return, address);
  assert.equal(successor, 0);
  const response: Message = { frame: { version: 1, kind: 'response', id: 'c:1', result: 42 } };
  seen[0]!.message.return!.wire.send([], response);
  assert.equal(reply.queue[0]?.message, response);
  releaseNext();
});

test('empty paths and keys remain distinct, and an empty mount can own an attachment', async () => {
  const child = new QueuedEndpoint();
  const mounted = mount(new Map([['', child]]));
  const paths: Path[] = [];
  const detach = mounted.receive({
    message: (path) => {
      paths.push(path);
    },
  });
  assert.throws(() => mounted.send([], event), { code: 'no_route' });
  assert.throws(() => mounted.send(['missing'], event), { code: 'no_route' });
  assert.throws(() => mounted.send(['', '\ud800'], event), { code: 'invalid_path' });
  mounted.send([''], event);
  mounted.send(['', ''], event);
  await child.drain();
  assert.deepEqual(paths, [[''], ['', '']]);
  detach();
  const empty = mount(new Map());
  let closed = 0;
  const release = empty.receive({
    closed: () => {
      closed++;
    },
  });
  assert.throws(() => empty.receive({}), { code: 'receiver_exists' });
  assert.throws(() => empty.send([], event), { code: 'no_route' });
  release();
  empty.receive({
    closed: () => {
      closed++;
    },
  });
  empty.close();
  empty.close();
  assert.equal(closed, 1);
});
